tests/test_butler.py

```
@@ -1193,5 +1202,117 @@ def tearDown(self):
+
+        This is equivalent to tempfile.mkdtemp as this is what self.root
+        becomes when useTempRoot is True.
+        """
+        rndstr = "".join(
+            random.choice(string.ascii_uppercase + string.digits) for _ in range(20)
+        )
+        return rndstr + "/"
+
+    @classmethod
+    def setUpClass(cls):
+        # Do the same as inherited class
+        cls.storageClassFactory = StorageClassFactory()
+        cls.storageClassFactory.addFromConfig(cls.configFile)
+
+        # Run a local webdav server on which tests will be run
+        t = Thread(target=cls._serveWebdav, daemon=True)
+        t.start()
+        # Wait for it to start
+        time.sleep(3)
+
+    def setUp(self):
+        os.environ["WEBDAV_AUTH_METHOD"] = "TOKEN"
+        os.environ["WEBDAV_BEARER_TOKEN"] = "XXXXXX"
+
+        config = Config(self.configFile)
+        uri = ButlerURI(config[".datastore.datastore.root"])
+        self.serverName = uri.netloc
+
+        if self.useTempRoot:
+            self.root = self.genRoot()
+        rooturi = f"http://{self.serverName}/{self.root}"
+        config.update({"datastore": {"datastore": {"root": rooturi}}})
+
+        self.datastoreStr = f"datastore={self.root}"
+        self.datastoreName = [f"WebdavDatastore@{rooturi}"]
+
+        Butler.makeRepo(rooturi, config=config, forceConfigRoot=False)
+        self.tmpConfigFile = posixpath.join(rooturi, "butler.yaml")
+
+    def tearDown(self):
+        rooturi = f"http://{self.serverName}/{self.root}"
+        uri = ButlerURI(rooturi)
+        uri.remove()
+
+    def _serveWebdav():
+        """Starts a local webdav-compatible HTTP server,
+        Listening on http://localhost:8080
+        This server only runs when this test class is instantiated,
+        and then shuts down. Must be started is a separate thread.
+        """
+        root_path = gettempdir()
+
+        config = {
+            "host": "0.0.0.0",
+            "port": 8080,
```

Do we have to use 8080? Does it deal with people who are running local web servers there already?  If you run the tests with `pytest -n5 tests/test_butler.py` won't it start multiple web servers, one for each subprocess? Don't they clash?

---

We don't necessarily have to use 8080, any port above 1024 is probably fine (we may run into permissions issues if using any port below 1024).
I've added some necessary checks : if there is another service running on the selected port, the webdav server will fail to start and all tests will raise an OSError and fail.

---

Running pytest with `-n5` for example does not seem to be an issue and succeeds. I'm not sure if it's something related to how `pytest` works with multithread, or if it's just the fact that, once the first webdav server is up, subsequent webdav server threads silently fail and tests all run against the first one.

---

I wonder if using port 0 works (it depends on whether you can ask the server what port it has been assigned).

---

It seems plausible that they are sharing the same server. This seems a bit risky if the first subprocess completes before the others have finished talking to it. I imagine we are lucking out because the webdav tests are a small fraction of all the tests.